### PR TITLE
rustc: add support for armv7l targets

### PR DESCRIPTION
**Pull request by urbas**
###### Motivation for this change
rustc and Rust packages would not compile on Raspbian on a Raspberry Pi. It turns out that Rust's ARMv7 targets are slightly different from Nix's.

Addresses issue #72473

###### Things done
This change adds a function that maps `armv7a` and `armv7l` architectures to `armv7`. The former are used in Raspbian on Raspberry Pi but are unknown to rustc. Rust knows only `armv7`.
    
In addition, this change also disables the `m_hugefile` test in `e2fsprogs` as it was failing on ARMv7 Raspbian on the Raspberry Pi.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @edolstra @madjar @cstrahan @globin @Havvy

### CODE REVIEW

File: pkgs/tools/filesystems/e2fsprogs/default.nix [hunk 1]
```diff
@@ -1,5 +1,21 @@
-{ stdenv, buildPackages, fetchurl, fetchpatch, pkgconfig, libuuid, gettext, texinfo }:
+{ stdenv, buildPackages, fetchurl, fetchpatch, pkgconfig, libuuid, gettext, texinfo, lib }:
```
Comment (Mic92) [inline pkgs/tools/filesystems/e2fsprogs/default.nix]: Could you make this a separate commit?

**Review by Mic92 — APPROVED** [review]
Have not tested, but code seems fine.

**Review by Gerschtli — APPROVED** [review]
Haven't tested it either, but diff LGTM.

**Review by matthewbauer — APPROVED** [review]

**Comment by doronbehar** [comment]
BTW [Rust 1.39.0 was released today](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1390-2019-11-07) so it might worth updating it here as well.

**Comment by edolstra** [comment]
That's already done in #72980.

**Comment by Mic92** [comment]
@GrahamcOfBorg build rustc

**Review by lopsided98 — CHANGES_REQUESTED** [review]
I wrote a similar patch a while ago to add support for armv7l and armv6l, see: https://github.com/lopsided98/nixpkgs/commit/d6d2c1c893532e93777bdd1d7b7333718e178af9. My patch fixes the target in a few more places that you missed, but I missed the one in build-crate.nix.

File: pkgs/development/compilers/rust/default.nix [hunk 2]
```diff
@@ -5,13 +5,22 @@
...
+  makeRustPlatform = { rustc, cargo, ... }: rec {
     rust = {
       inherit rustc cargo;
     };
 
+    toRustTarget = nixTarget: {
+        x86_64-pc-mingw32 = "x86_64-pc-windows-gnu";
+      }.${nixTarget} or (
+        let
+          armMatch = builtins.match "armv7[al]-(.*)" nixTarget;
```
[the remark this answers is not in the document]
Comment (lopsided98) [reply pkgs/development/compilers/rust/default.nix]: Something like this: https://github.com/lopsided98/nixpkgs/commit/9015731404369317b403d97d4f42d796c3c297d0#diff-4100057d032c0d790107c2739448233cR8

File: pkgs/tools/filesystems/e2fsprogs/default.nix [hunk 3]
```diff
@@ -1,5 +1,21 @@
...
+    "m_image_mmp"
+    "m_mmp"
+    "m_mmp_bad_csum"
+    "m_mmp_bad_magic"
+    "t_mmp_1on"
+    "t_mmp_2off"
+  ] ++
+    # Remove a test that fails on 32bit ARMv7 architectures
+    # https://github.com/NixOS/nixpkgs/issues/72473
+    (lib.optional (stdenv.hostPlatform.parsed.cpu.name == "armv7l") "m_hugefile");
```
[the remark this answers is not in the document]
Comment (urbas) [reply pkgs/tools/filesystems/e2fsprogs/default.nix]: Tbh, it feels wrong to disable tests just because of Raspbian. If it works on NixOS, I'd rather keep the test enabled. Besides, this commit doesn't technically fit the title of the pull request. I'll remove this commit.

File: pkgs/development/compilers/rust/default.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (urbas) [reply pkgs/development/compilers/rust/default.nix]: My fixes look extremely similar to yours. I also see they well predate mine. I am sorry, I wasn't aware of them.

Btw, I also agree with your suggestions. I can fix this PR. However, I can also retract this pull request so you can go ahead and create yours. Let me know which way you'd prefer.

[the remark this answers is not in the document]
Comment (FRidh) [reply pkgs/development/compilers/rust/default.nix]: How will you continue?

[the remark this answers is not in the document]
Comment (urbas) [reply pkgs/development/compilers/rust/default.nix]: I'll apply the fixes @lopsided98 recommended.

[the remark this answers is not in the document]
Comment (urbas) [reply pkgs/development/compilers/rust/default.nix]: I now applied the fixes.

**Review by Gerschtli — APPROVED** [review]

**Comment by lopsided98** [comment]
Sorry for not responding, but I can submit my own version of the PR if you want. I see that you integrated most of it already, but I noticed a couple of things still missing (in `.cargo/config` and bootstrapping hashes).

Also, we put our mapping functions inside different attrsets (`rust` for me vs. `rustPlatform` for you). This doesn't much functional difference, but it might be good to get the opinion of a Rust maintainer about where it belongs.

**Comment by lopsided98** [comment]
I submitted it: #73472

**Comment by urbas** [comment]
Yes, I intentionally left some things out because they caused errors in CI.
I recommend we attack this problem in 2 steps. This PR goes in first (since
it has already been reviewed and approved) and then the second PR (where
you'll put in the improvements) can be much smaller and easier to review.
Does that sound reasonable?

**Comment by lopsided98** [comment]
This PR has the same problem, ofborg just doesn't catch it. On x86_64 macOS, the `toRustTarget` function returns `x86_64-apple-darwin-unknown`, which will probably cause the rustc build to fail because that is not a target triple Rust knows about.

**Comment by lopsided98** [comment]
Here is my updated mapping function: https://github.com/NixOS/nixpkgs/pull/73472/files#diff-4100057d032c0d790107c2739448233cR14

It is not the most elegant, but it works for all the platforms that we have bootstrap tarballs for, as well as MinGW (which actually did not need a special case) and armv7a.

**Comment by urbas** [comment]
@lopsided98: Sounds much better.

I'm throwing this PR out in favour of #73472
